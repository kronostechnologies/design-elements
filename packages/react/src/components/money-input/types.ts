import { InputFieldProps } from '../input/types';

type CommonInputFieldProps = Pick<InputFieldProps,
    | 'className'
    | 'required'
    | 'disabled'
    | 'label'
    | 'hint'
>;

export interface MoneyInputProps extends CommonInputFieldProps {
    value?: number | null;
    /**
     * Sets input locale and changes visual format accordingly
     * @default fr-CA
     */
    locale?: string;
    /**
     * Sets currency
     * @default CAD
     */
    currency?: string;
    /**
     * Sets number of decimals
     * @default 2
     */
    precision?: number;
    validationErrorMessage?: string;

    onChange?(value: number | null, formattedValue: string): void;
}
