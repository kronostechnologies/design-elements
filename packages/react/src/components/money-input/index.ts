export { MoneyInput, type MoneyInputProps } from './money-input';
