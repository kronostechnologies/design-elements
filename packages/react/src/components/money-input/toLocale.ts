export function toLocale(
    language: string,
    currency: string,
): string {
    if (!language.match(/^[a-z]{2}-[a-z]{2}$/i)) {
        switch (currency.toUpperCase()) {
            case 'CAD':
                return `${language}-CA`;
            case 'USD':
                return `${language}-US`;
        }
    }

    return language;
}
