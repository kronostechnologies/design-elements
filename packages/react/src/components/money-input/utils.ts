import { formatCurrency } from '../../utils/currency';

export function safeFormatCurrency(
    value: number | null,
    precision: number,
    locale: string,
    currency: string,
): string {
    return value === null ? '' : formatCurrency(value, precision, locale, currency);
}

export function roundValueToPrecision(amount: number | null, precision: number): number | null {
    return amount !== null && !Number.isNaN(amount)
        ? Math.round(amount * (10 ** precision)) / (10 ** precision)
        : null;
}

export function parseAndRound(val: string, precision: number): number | null {
    return val === '' ? null : roundValueToPrecision(Number(val.replace(',', '.')), precision);
}
