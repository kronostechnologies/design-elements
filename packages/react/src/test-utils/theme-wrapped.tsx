import React, { ComponentProps, ComponentType, ReactElement } from 'react';
import { equisoftTheme, ThemeWrapper } from '..';

type ThemeProps = Pick<ComponentProps<typeof ThemeWrapper>, 'children'>;

export function themeProvider(): ComponentType {
    return  ThemeProvider as ComponentType;
}

export function ThemeProvider({ children }: ThemeProps): ReactElement {
    return (
        <ThemeWrapper theme={equisoftTheme}>
            {children}
        </ThemeWrapper>
    );
}

export function ThemeWrapped(children: ReactElement): ReactElement {
    return (
        <ThemeWrapper theme={equisoftTheme}>
            {children}
        </ThemeWrapper>
    );
}
