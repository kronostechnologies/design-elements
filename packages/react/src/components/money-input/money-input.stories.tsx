import React from 'react';
import { Button } from '../buttons/button';
import { MoneyInput } from './money-input';

export default {
    title: 'Money Input',
    component: MoneyInput,
};

export const normal = () => (
    <>
        <MoneyInput label="Entrez un montant"/>
        <MoneyInput label="Choose a number" locale="en-CA"/>
    </>
);

export const englishLocale = () => (
    <MoneyInput label="Choose a number" locale="en-CA"/>
);
export const noLabel = () => (
    <MoneyInput/>
);
export const disabled = () => (
    <MoneyInput disabled label="Entrez un montant"/>
);
export const withCurrency = () => (
    <MoneyInput label="Entrez un montant" currency="USD"/>
);
export const withPrecision = () => (
    <MoneyInput label="Entrez un montant" precision={0}/>
);
export const withValue = () => (
    <MoneyInput label="Entrez un montant" value={350}/>
);
export const required = () => (
    <form onSubmit={(event) => event.preventDefault()}>
        <MoneyInput required label="Entrez un montant"/>
        <Button buttonType="primary" type="submit">Soumettre</Button>
    </form>
);
export const customErrorMessage = () => (
    <form onSubmit={(event) => event.preventDefault()}>
        <MoneyInput required label="Entrez un montant" validationErrorMessage="Custom error message."/>
    </form>
);
export const onChangeCallback = () => (
    <MoneyInput
        label="Entrez un montant"
        onChange={(value, formattedValue) => {
            console.log('value:', value);
            console.log('formattedValue: ', formattedValue);
        }}
    />
);
