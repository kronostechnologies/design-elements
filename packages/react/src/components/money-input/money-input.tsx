import { ChangeEvent, ReactElement, useCallback, useEffect, useRef, useState } from 'react';
import styled from 'styled-components';
import { useTranslation } from '../../i18n/use-translation';
import { formatCurrency } from '../../utils/currency';
import { TextInput } from '../text-input/text-input';

type Language = 'en' | 'fr';

const InputWrapper = styled.div<{ language: Language }>`
    input {
        text-align: ${({ language }) => (language === 'en' ? 'left' : 'right')};
        transition: width 300ms;
        width: 132px;
    }
`;

function safeFormatCurrency(
    value: number | null = null,
    precision: number,
    locale: string,
    currency: string,
): string {
    return value === null ? '' : formatCurrency(value, precision, locale, currency);
}

interface Props {
    className?: string;
    disabled?: boolean;
    required?: boolean;
    /**
     * Message displayed in case of validation error
     * @default Invalid number.
     */
    validationErrorMessage?: string;
    label?: string;
    value?: number | null;
    /**
     * Sets input locale and changes visual format accordingly
     * @default fr-CA
     */
    locale?: string;
    /**
     * Sets currency
     * @default CAD
     */
    currency?: string;
    /**
     * Sets number of decimals
     * @default 2
     */
    precision?: number;
    hint?: string;

    onChange?(value: number | null, formattedValue: string): void;
}

function roundValueToPrecision(amount: number | null, precision: number): number | null {
    return amount !== null && !Number.isNaN(amount)
        ? Math.round(amount * (10 ** precision)) / (10 ** precision)
        : null;
}

function parseAndRound(val: string, precision: number): number | null {
    return val === '' ? null : roundValueToPrecision(Number(val.replace(',', '.')), precision);
}

export function MoneyInput({
    className,
    required,
    disabled,
    label,
    onChange,
    precision = 2,
    value = null,
    validationErrorMessage,
    locale = 'fr-CA',
    currency = 'CAD',
    hint,
}: Props): ReactElement {
    const { t } = useTranslation('money-input');
    const inputElement = useRef<HTMLInputElement>(null);
    const language: Language = locale.split('-')[0] as Language;
    const [displayValue, setDisplayValue] = useState(safeFormatCurrency(value, precision, locale, currency));
    const [maskedValue, setMaskedValue] = useState(safeFormatCurrency(value, precision, locale, currency));
    const [hasFocus, setHasFocus] = useState<boolean>(false);

    const updateFormattedValue: (rawValue: string) => void = useCallback((rawValue) => {
        const roundedValue = parseAndRound(rawValue, precision);
        const newMaskedValue: string = safeFormatCurrency(roundedValue, precision, locale, currency);

        setDisplayValue(newMaskedValue);
        if (maskedValue !== newMaskedValue) {
            setMaskedValue(newMaskedValue);
            if (onChange) {
                onChange(roundedValue, newMaskedValue);
            }
        }
    }, [currency, locale, maskedValue, onChange, precision]);

    useEffect(() => {
        if (inputElement.current != null) {
            if (!hasFocus && displayValue !== maskedValue) {
                updateFormattedValue(inputElement.current.value);
            }
        }
    }, [hasFocus, displayValue, maskedValue, updateFormattedValue]);

    useEffect(() => {
        if (hasFocus) {
            inputElement.current?.select();
        }
    }, [hasFocus]);

    useEffect(() => {
        const newValue = safeFormatCurrency(value, precision, locale, currency);
        setDisplayValue(newValue);
        setMaskedValue(newValue);
    }, [currency, locale, precision, value]);

    const handleBlurEvent: () => void = useCallback(() => {
        setHasFocus(false);
    }, [setHasFocus]);

    const handleFocusEvent: () => void = useCallback(() => {
        const roundedValue = roundValueToPrecision(value, precision);
        const roundedValueAsString = roundedValue === null ? '' : roundedValue.toString();
        setDisplayValue(roundedValueAsString);
        setHasFocus(true);
    }, [setHasFocus, precision, value]);

    const handleChangeEvent: (event: ChangeEvent<HTMLInputElement>) => void = useCallback((event) => {
        const rawValue = event.target.value;
        const mask = /[^0-9.,]/g;
        const nextDisplayValue = rawValue.replace(mask, '');

        event.preventDefault();
        setDisplayValue(nextDisplayValue);
    }, []);

    return (
        <InputWrapper language={language}>
            <TextInput
                className={className}
                required={required}
                disabled={disabled}
                ref={inputElement}
                type="text"
                inputMode="numeric"
                value={displayValue}
                label={label}
                placeholder="$"
                onChange={handleChangeEvent}
                onBlur={handleBlurEvent}
                onFocus={handleFocusEvent}
                validationErrorMessage={validationErrorMessage || t('validationErrorMessage')}
                hint={hint}
            />
        </InputWrapper>
    );
}
