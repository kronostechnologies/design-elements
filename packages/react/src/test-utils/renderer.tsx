import { ThemeWrapped } from '@design-elements/test-utils/theme-wrapped';
import { mount, ReactWrapper, render } from 'enzyme';
import React, { Component, ReactElement } from 'react';
import { MemoryRouter } from 'react-router-dom';
import { DeviceContextProvider } from '../components/device-context-provider/device-context-provider';

export function mountWithProviders<C extends Component, P = C['props'], S = C['state']>(
    component: ReactElement<P>,
): ReactWrapper<P, S, C> {
    return mount(component, {
        wrappingComponent: AllProviders,
    });
}

export function renderWithProviders(
    component: ReactElement,
): Cheerio {
    return render(<AllProviders>{component}</AllProviders>);
}

function AllProviders({ children }: { children: ReactElement }): ReactElement {
    return (
        <MemoryRouter>
            <DeviceContextProvider>
                {ThemeWrapped(children)}
            </DeviceContextProvider>
        </MemoryRouter>
    );
}
