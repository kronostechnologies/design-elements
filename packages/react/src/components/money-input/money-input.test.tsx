import { fireEvent, render } from '@testing-library/react';
import React from 'react';
import renderer from 'react-test-renderer';
import { themeProvider, ThemeWrapped } from '../../test-utils/theme-wrapped';
import { MoneyInput } from './money-input';
jest.mock('uuid/v4');

describe('CurrencyInput Component', () => {
    it('should remove formatting on focus', () => {
        const { container } = render(<MoneyInput precision={2} value={12345.25} />, { wrapper: themeProvider() });
        const input = getInputElement(container);

        fireEvent.focus(input);

        expect(input.value).toEqual('12345.25');
    });

    it('should format on blur', () => {
        const { container } = render(
            <MoneyInput precision={2} value={12345.25} />, { wrapper: themeProvider() },
        );
        const input = getInputElement(container);

        fireEvent.focus(input);
        fireEvent.blur(input);

        expect(input.value).toMatchFormattedMoney('12 345,25 $');
    });

    it('should remove fractions when precision is 0 and changing value', () => {
        const { container } = render(<MoneyInput precision={0} value={12345.25} />, { wrapper: themeProvider() });
        const input = getInputElement(container);

        fireEvent.focus(input);
        fireEvent.change(input, { target: { value: '12345' } });
        fireEvent.blur(input);

        expect(input.value).toMatchFormattedMoney('12 345 $');
    });

    it('should use precision when changing value', () => {
        const { container } = render(<MoneyInput precision={2} value={null} />, { wrapper: themeProvider() });
        const input = getInputElement(container);

        fireEvent.focus(input);
        fireEvent.change(input, { target: { value: '123.457' } });
        fireEvent.blur(input);

        expect(input.value).toMatchFormattedMoney('123,46 $');
    });

    it('should format according to locale when changing value', () => {
        const { container } = render(<MoneyInput precision={2} language="en" />, { wrapper: themeProvider() });
        const input = getInputElement(container);

        fireEvent.focus(input);
        fireEvent.change(input, { target: { value: '12345' } });
        fireEvent.blur(input);

        expect(input.value).toMatchFormattedMoney('$12,345.00');
    });

    it('should format provided value', () => {
        const { container } = render(<MoneyInput precision={2} value={12345.25} />, { wrapper: themeProvider() });
        const input = getInputElement(container);

        expect(input.value).toMatchFormattedMoney('12 345,25 $');
    });

    it('should select all text on focus', () => {
        const { container } = render(<MoneyInput precision={2} value={12345} />, { wrapper: themeProvider() });
        const input = getInputElement(container);

        fireEvent.focus(input);

        expect(input.selectionStart).toEqual(0);
        expect(input.selectionEnd).toEqual(5);
    });

    it('should respect fractions when calling change handler', () => {
        const handleChange = jest.fn();
        const { container } = render(<MoneyInput language="en" precision={0} value={0} onChange={handleChange} />,
            { wrapper: themeProvider() });
        const input = getInputElement(container);

        fireEvent.focus(input);
        fireEvent.change(input, { target: { value: 12.345 } });
        fireEvent.blur(input);

        expect(input.value).toMatchFormattedMoney('$12');
        expect(handleChange).toBeCalledWith(12, '$12');
    });

    it('should allow 0 as a value', () => {
        const { container } = render(<MoneyInput precision={0} value={0} />, { wrapper: themeProvider() });
        const input = getInputElement(container);

        fireEvent.focus(input);
        fireEvent.blur(input);

        expect(input.value).toMatchFormattedMoney('0 $');
    });

    it('should allow to be empty', () => {
        const { container } = render(<MoneyInput precision={0} value={0} />, { wrapper: themeProvider() });
        const input = getInputElement(container);

        fireEvent.focus(input);
        fireEvent.change(input, { target: { value: '' } });
        fireEvent.blur(input);

        expect(input.value).toMatchFormattedMoney('');
    });

    it('matches snapshot (fr)', () => {
        const tree = renderer.create(
            ThemeWrapped(<MoneyInput value={100} />),
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

    it('matches snapshot (en)', () => {
        const tree = renderer.create(
            ThemeWrapped(<MoneyInput value={100} language="en" />),
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });

});

function getInputElement(container: HTMLElement): HTMLInputElement {
    // tslint:disable-next-line: no-unnecessary-type-assertion
    return container.querySelector('input') as HTMLInputElement;
}
