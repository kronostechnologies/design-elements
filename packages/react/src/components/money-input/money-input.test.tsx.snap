// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrencyInput Component matches snapshot (en-CA) 1`] = `
.c1 {
  margin: 0 0 var(--spacing-3x);
}

.c1 input,
.c1 select,
.c1 textarea {
  border-color: #60666E;
}

.c1 > :nth-child(0) {
  margin-bottom: var(--spacing-half);
}

.c2 {
  background: #FFFFFF;
  border: 1px solid #60666E;
  border-radius: var(--border-radius);
  box-sizing: border-box;
  color: #000000;
  font-family: inherit;
  font-size: 0.875rem;
  -webkit-letter-spacing: 0.015rem;
  -moz-letter-spacing: 0.015rem;
  -ms-letter-spacing: 0.015rem;
  letter-spacing: 0.015rem;
  line-height: 1.5rem;
  margin: 0;
  min-height: var(--size-2x);
  outline: none;
  padding: 0 var(--spacing-1x);
  width: 100%;
}

.c2::-webkit-input-placeholder {
  color: #60666E;
}

.c2::-moz-placeholder {
  color: #60666E;
}

.c2:-ms-input-placeholder {
  color: #60666E;
}

.c2::placeholder {
  color: #60666E;
}

.c2:disabled {
  background-color: #F1F2F2;
  border-color: #DBDEE1;
  color: #DBDEE1;
}

.c2:disabled,
.c2:disabled::-webkit-input-placeholder {
  color: #B7BBC2;
}

.c2:disabled,
.c2:disabled::-moz-placeholder {
  color: #B7BBC2;
}

.c2:disabled,
.c2:disabled:-ms-input-placeholder {
  color: #B7BBC2;
}

.c2:disabled,
.c2:disabled::placeholder {
  color: #B7BBC2;
}

.c2:not(:focus) {
  -webkit-transition: all .25s ease-in-out;
  transition: all .25s ease-in-out;
  outline: 2px solid transparent;
  outline-offset: -2px;
}

.c2:focus {
  -webkit-transition: all .25s ease-in-out;
  transition: all .25s ease-in-out;
  box-shadow: 0 0 0 2px #006296;
  outline: 2px solid #84C6EA;
  outline-offset: -2px;
}

.c0 input {
  text-align: left;
  width: 132px;
}

<div
  class="c0"
>
  <div
    class="c1"
    data-testid="field-container"
  >
    <input
      class="c2"
      data-testid="text-input"
      id="uuid1"
      inputmode="numeric"
      placeholder="$"
      type="text"
      value="$100.00"
    />
  </div>
</div>
`;

exports[`CurrencyInput Component matches snapshot (en-US) 1`] = `
.c1 {
  margin: 0 0 var(--spacing-3x);
}

.c1 input,
.c1 select,
.c1 textarea {
  border-color: #60666E;
}

.c1 > :nth-child(0) {
  margin-bottom: var(--spacing-half);
}

.c2 {
  background: #FFFFFF;
  border: 1px solid #60666E;
  border-radius: var(--border-radius);
  box-sizing: border-box;
  color: #000000;
  font-family: inherit;
  font-size: 0.875rem;
  -webkit-letter-spacing: 0.015rem;
  -moz-letter-spacing: 0.015rem;
  -ms-letter-spacing: 0.015rem;
  letter-spacing: 0.015rem;
  line-height: 1.5rem;
  margin: 0;
  min-height: var(--size-2x);
  outline: none;
  padding: 0 var(--spacing-1x);
  width: 100%;
}

.c2::-webkit-input-placeholder {
  color: #60666E;
}

.c2::-moz-placeholder {
  color: #60666E;
}

.c2:-ms-input-placeholder {
  color: #60666E;
}

.c2::placeholder {
  color: #60666E;
}

.c2:disabled {
  background-color: #F1F2F2;
  border-color: #DBDEE1;
  color: #DBDEE1;
}

.c2:disabled,
.c2:disabled::-webkit-input-placeholder {
  color: #B7BBC2;
}

.c2:disabled,
.c2:disabled::-moz-placeholder {
  color: #B7BBC2;
}

.c2:disabled,
.c2:disabled:-ms-input-placeholder {
  color: #B7BBC2;
}

.c2:disabled,
.c2:disabled::placeholder {
  color: #B7BBC2;
}

.c2:not(:focus) {
  -webkit-transition: all .25s ease-in-out;
  transition: all .25s ease-in-out;
  outline: 2px solid transparent;
  outline-offset: -2px;
}

.c2:focus {
  -webkit-transition: all .25s ease-in-out;
  transition: all .25s ease-in-out;
  box-shadow: 0 0 0 2px #006296;
  outline: 2px solid #84C6EA;
  outline-offset: -2px;
}

.c0 input {
  text-align: left;
  width: 132px;
}

<div
  class="c0"
>
  <div
    class="c1"
    data-testid="field-container"
  >
    <input
      class="c2"
      data-testid="text-input"
      id="uuid1"
      inputmode="numeric"
      placeholder="$"
      type="text"
      value="$100.00"
    />
  </div>
</div>
`;

exports[`CurrencyInput Component matches snapshot (fr-CA) 1`] = `
.c1 {
  margin: 0 0 var(--spacing-3x);
}

.c1 input,
.c1 select,
.c1 textarea {
  border-color: #60666E;
}

.c1 > :nth-child(0) {
  margin-bottom: var(--spacing-half);
}

.c2 {
  background: #FFFFFF;
  border: 1px solid #60666E;
  border-radius: var(--border-radius);
  box-sizing: border-box;
  color: #000000;
  font-family: inherit;
  font-size: 0.875rem;
  -webkit-letter-spacing: 0.015rem;
  -moz-letter-spacing: 0.015rem;
  -ms-letter-spacing: 0.015rem;
  letter-spacing: 0.015rem;
  line-height: 1.5rem;
  margin: 0;
  min-height: var(--size-2x);
  outline: none;
  padding: 0 var(--spacing-1x);
  width: 100%;
}

.c2::-webkit-input-placeholder {
  color: #60666E;
}

.c2::-moz-placeholder {
  color: #60666E;
}

.c2:-ms-input-placeholder {
  color: #60666E;
}

.c2::placeholder {
  color: #60666E;
}

.c2:disabled {
  background-color: #F1F2F2;
  border-color: #DBDEE1;
  color: #DBDEE1;
}

.c2:disabled,
.c2:disabled::-webkit-input-placeholder {
  color: #B7BBC2;
}

.c2:disabled,
.c2:disabled::-moz-placeholder {
  color: #B7BBC2;
}

.c2:disabled,
.c2:disabled:-ms-input-placeholder {
  color: #B7BBC2;
}

.c2:disabled,
.c2:disabled::placeholder {
  color: #B7BBC2;
}

.c2:not(:focus) {
  -webkit-transition: all .25s ease-in-out;
  transition: all .25s ease-in-out;
  outline: 2px solid transparent;
  outline-offset: -2px;
}

.c2:focus {
  -webkit-transition: all .25s ease-in-out;
  transition: all .25s ease-in-out;
  box-shadow: 0 0 0 2px #006296;
  outline: 2px solid #84C6EA;
  outline-offset: -2px;
}

.c0 input {
  text-align: right;
  width: 132px;
}

<div
  class="c0"
>
  <div
    class="c1"
    data-testid="field-container"
  >
    <input
      class="c2"
      data-testid="text-input"
      id="uuid1"
      inputmode="numeric"
      placeholder="$"
      type="text"
      value="100,00Â $"
    />
  </div>
</div>
`;
